### DISCUSSION TITLE
does anyone know why this isn't working

### DISCUSSION THREAD

**Discussion by AlexdaCat** [FabricMC/fabric-api#1848] [Mod Dev Support] [answered]
public class MyBlock extends Block {
	public static final BooleanProperty BLUEPOWERED = BooleanProperty.of("bluepowered");

	@Override
	protected void appendProperties(StateManager.Builder<Block, BlockState> stateManager) {
		stateManager.add(BLUEPOWERED);
	}

	public MyBlock(Settings settings) {
		super(settings);
		setDefaultState(getStateManager().getDefaultState().with(BLUEPOWERED, false));
	}
}

by the way the file path is `src-main-java-alexdacat-moreredstone-fabric-moreredstone.java`
should it be in one of the block;s files?
(moreredstone fabric is the name of the mod)

**Comment by AlexdaCat** [comment]
also is there any easy way to have an nbt tag/ blockstate be able to be applied to any block in the game or mods?
kinda pretty important for what im going to be doing

**Comment by AlexdaCat** [comment]
I _might_ be stupid

**Comment by warjort** [comment, marked as answer]
````does anyone know why this isn't working````

Define "not working".
You can see the block states for a block if you press F3 (the debug overlay) and look at the block.

````also is there any easy way to have an nbt tag/ blockstate be able to be applied to any block in the game or mods?````

Not really.

For block states you could use a mixin to intercept the generic net.minecraft.block.(Abstract)Block methods, but changing the states for a block could interact badly with what the block is doing, e.g. its blockstates.json or dynamic rendering
The block also wouldn't know how to preserve your state when it changes its known states.

If a block is a BlockEntity it can have NBT. 
However, this is only used for serialization (saving to file or passing over the network).
At runtime each block decides what the NBT means and doesn't have to use the NBT directly.

Summary;
I don't think you want to go down this route. It will be very complicated and very brittle to changes either in Minecraft or other mods.

**Reply by AlexdaCat** [reply]
well i was thinking of a cool idea on how redstone works cuz bluestone could have bluehardpowering and bluesoftpowering so even through blocks it wouldnt interact with redstone

**Reply by AlexdaCat** [reply]
actually, I could just copy how redstone does hard/soft powering stuff, right?
and just change it?

**Reply by AlexdaCat** [reply]
I was talking about it not showing correctly and now I realize I didn't really say that but oh well because I shouldn't mess with blockstates yet anyways

**Reply by warjort** [reply]
Its an idea that has been done before so it is possible.

These are what I found using a quick search, but IIRC the idea is older than these.
https://www.blockminecraft.com/blue-power-mod/
https://www.curseforge.com/minecraft/mc-mods/mumbos-redstone-additions
https://www.curseforge.com/minecraft/mc-mods/colored-redstone

The above are all forge mods, I don't know if somebody has done it for fabric?

**Comment by warjort** [comment]
````should it be in one of the block;s files?````

Do you mean you want to change the rendering based on the state?

This is done in the blockstates json file by choosing different models for different states, e.g. simple way

````
{
  "variants": {
    "bluepowered=false": {
      "model": "moreredstone:block/bluepowered_false"
    },
    "bluepowered=true": {
      "model": "moreredstone:block/bluepowered_true"
    }
}
````

From what you are doing, you probably want to look at look at minecraft's
````assets/minecraft/blockstates/redstone_wire.json````
That uses the alternative "multipart" definition for blockstates -> models.

**Comment by Daomephsta** [comment]
> also is there any easy way to have an nbt tag/ blockstate be able to be applied to any block in the game or mods? kinda pretty important for what im going to be doing

All possible blockstates are computed during startup and cached. Adding a boolean property to all blocks would *double* the number of blockstates, and thus double the RAM used to store the blockstate cache. 
Blockstates are a useful way of storing information, but you need to be aware of how many total blockstates your block has. Adding properties to existing blocks requires even more care, particularly if you're adding properties to a lot of blocks.

**Comment by warjort** [comment]
This is an example mod that plays around with states for other blocks:
https://github.com/Virtuoel/Towelette
It allows any non solid block to be "waterlogged" and for the fluid to be anything, not just water.

Here, you can get a glimpse of how much work is involved (and this is just for the vanilla non-solid blocks):
https://github.com/Virtuoel/Towelette/tree/fabric/1.14.4-1.x.x/src/main/java/virtuoel/towelette/mixin
